src/main/java/it/gov/pagopa/xpay/controller/XPayController.java

```diff
@@ -0,0 +1,146 @@
+            "   </body>" +
+            "</html>";
+
+    @PostMapping("/ecomm/api/paga/autenticazione3DS")
+    public ResponseEntity<Object> paymentAuthorization(@RequestBody XPayAuthRequest request) {
+
+        log.info("Create AuthResponse for transactioId: " + request.getCodiceTransazione());
+
+        Long timeStamp =System.currentTimeMillis();
+        String idOperazione = createIdOperazione();
+
+        try {
+            String macToReturn = getMacToReturn(request.getCodiceTransazione(), request.getDivisa(), request.getImporto(), request.getTimeStamp());
+
+            saveXpayOnDb(idOperazione, request, timeStamp);
+
+            if (request.getMac().equals(macToReturn)) {
+               return createOkResponse(timeStamp, macToReturn,idOperazione);
+
+            } else {
+                XpayError xpayError = new XpayError(3L, "MAC errato");
+                return createKoResponse(timeStamp, xpayError, idOperazione);
+
+            }
+        } catch (Exception e) {
+            XpayError xpayError = new XpayError(50L, "Impossibile calcolare il mac");
+            return createKoResponse(timeStamp, xpayError, idOperazione);
+        }
+
+    }
+
+    private void saveXpayOnDb(String idOperazione, XPayAuthRequest request, Long timeStamp) {
+        XPayPayment xPayPayment = new XPayPayment();
+
+        xPayPayment.setIdOperazione(idOperazione);
+        xPayPayment.setApiKey(request.getApiKey());
+        xPayPayment.setCvv(request.getCvv());
+        xPayPayment.setDivisa(request.getDivisa());
+        xPayPayment.setPan(request.getPan());
+        xPayPayment.setImporto(request.getImporto());
+        xPayPayment.setMac(request.getMac());
+        xPayPayment.setScadenza(request.getScadenza());
+        xPayPayment.setCodiceTransazione(request.getCodiceTransazione());
+        xPayPayment.setTimeStamp_request(request.getTimeStamp());
+        xPayPayment.setTimeStamp_response(timeStamp);
+        xPayPayment.setUrlRisposta(request.getUrlRisposta());
+
+
+        xPayRepository.save(xPayPayment);
+
+    }
+
+    private ResponseEntity<Object> createOkResponse(Long timeStamp, String mac, String idOperazione) {
+        XPayAuthResponse xPayAuthResponse = new XPayAuthResponse();
+        xPayAuthResponse.setEsito(EsitoXpay.OK);
+        xPayAuthResponse.setIdOperazione(idOperazione);
+        xPayAuthResponse.setHtml(htmlToReturn);
+        xPayAuthResponse.setTimeStamp(timeStamp);
+        xPayAuthResponse.setMac(mac);
+
+        return ResponseEntity.ok().body(xPayAuthResponse);
+
+    }
+
+    private ResponseEntity<Object> createKoResponse(Long timeStamp, XpayError xpayError, String idOperazione) {
+        XPayAuthResponse xPayAuthResponse = new XPayAuthResponse();
+        xPayAuthResponse.setEsito(EsitoXpay.KO);
+        xPayAuthResponse.setIdOperazione(idOperazione);
+        xPayAuthResponse.setTimeStamp(timeStamp);
+        xPayAuthResponse.setErrore(xpayError);
+
```

Comment:
@DavideAmatoNexi you are missing parameter mac: it should be returned also in KO response https://ecommerce.nexi.it/specifiche-tecniche/servertoserver/pagamento3dsecure/autenticazione.html